pkg/client/client.go

```diff
@@ -0,0 +1,330 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// ObjectKey identifies a Kubernetes Object.
+type ObjectKey types.NamespacedName
+
+// Client uses methods defined in runtime.Object to determine the REST endpoint
+// for Kubernetes resources.
+
+// Interface defines the capability for a Kubernetes client.
+type Interface interface {
+	// Create saves the object obj in the Kubernetes cluster.
+	Create(ctx context.Context, obj runtime.Object) error
+
+	// Get retrieves an obj for the given object key from the Kubernetes Cluster.
+	// obj must be a struct pointer so that obj can be updated with the response
+	// returned by the Server.
+	Get(ctx context.Context, key ObjectKey, obj runtime.Object) error
+
+	// Update updates the given obj in the Kubernetes cluster. obj must be a
+	// struct pointer so that obj can be updated with the content returned by the Server.
+	Update(ctx context.Context, obj runtime.Object) error
+
+	// Delete deletes the given obj from Kubernetes cluster.
+	Delete(ctx context.Context, obj runtime.Object) error
+
+	// List retrieves list of objects for a given namespace and list options. On a
+	// successful call, Items field in the list will be populated with the
+	// result returned from the server.
+	List(ctx context.Context, namespace string, opts metav1.ListOptions, list runtime.Object) error
+}
+
+// NewForConfig returns a Kubernetes client implementation given a
+// configuration.
+func NewForConfig(config *rest.Config) (Interface, error) {
+	if config == nil {
+		return nil, fmt.Errorf("missing config")
+	}
+	config = rest.CopyConfig(config)
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	cachedDiscoveryClient := cached.NewMemCacheClient(cs.Discovery())
+	restMapper := restmapper.NewDeferredDiscoveryRESTMapper(cachedDiscoveryClient)
+	restMapper.Reset()
+
+	dc, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &client{
+		conf:       config,
+		dc:         dc,
+		scheme:     scheme.Scheme,
```

Comment:
The standard client-go scheme is used over here.
Is it assumed that any custom types defined elsewhere need to add their types to the client-go scheme before the dynamic-client can correctly map the custom type to its GVK?
https://github.com/kubernetes-sigs/kubebuilder/blob/master/samples/controller/pkg/apis/samplecontroller/v1alpha1/register.go#L42

```Go
import (
	samplecontrollerv1alpha1 "github.com/kubernetes-sigs/kubebuilder/samples/controller/pkg/apis/samplecontroller/v1alpha1"
	"k8s.io/client-go/kubernetes/scheme"
)

func init() {
	samplecontrollerv1alpha1.AddToScheme(scheme.Scheme)
}

```

Reply:
> Is it assumed that any custom types defined elsewhere need to add their types to the client-go scheme before the dynamic-client can correctly map the custom type to its GVK?

No, would like to avoid that. I am thinking more along the lines of providing a hook to register the custom-type to the scheme of the client. will be trying that out today.